src/test/scala/akka/stream/contrib/RetrySpec.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.scaladsl._
+import akka.stream.testkit.scaladsl.{ TestSink, TestSource }
+import scala.util.{ Try, Failure, Success }
+
+class RetrySpecAutoFusingOn extends { val autoFusing = true } with RetrySpec
+class RetrySpecAutoFusingOff extends { val autoFusing = false } with RetrySpec
+
+trait RetrySpec extends BaseStreamSpec {
+
+  val failedElem: Try[Int] = Failure(new Exception("cooked failure"))
+  val flow = Flow[(Int, Int)].map {
+    case (i, j) if i % 2 == 0 => (failedElem, j)
+    case (i, j)               => (Success(i + 1), j)
+  }
+
+  "Retry" should {
+    "retry ints according to their parity" in {
```

Comment:
more tests are needed, especially around completion, and also testing deadlock-freedom.